Remove 7zip from compression module

johnwparent opened the pull request:
A hard dependency on 7zip was introduced on the Windows side to support a lack of the `lzma` module on earlier versions of Python. 7zip was also introduced as a Spack package, but could not be used as such initially as Clingo requires lzma support for extraction, and thus nothing could be installed if bootstrapping was required without 7zip existing in the PATH already (or having been externally detected).

We want to have as minimal a set of startup dependencies to Spack on Windows at first as humanly possible (ideally just git, Python, and MSVC/other compilers as we add support) so requiring 7zip is far from ideal. 

Now that we've dropped support for these older version of Python, we can consistently rely on the `lzma` Python module existing to extract Clingo for installation. 

This PR:

* removes references to 7zip in the compression module
* adds preference for Python's `lzma` module, with fallback to `xz` system util a la gzip, bzip2, and every other extension but `tar` and `zip`, which prefer system utils as the Python modules have significant issues.
* `Z` compressed archives are now only uncompressible via gzip (Windows is now the same as all other platforms were previously in that regard)
* General cleanup in the compression module

johnwparent committed:
Remove hard dependency on 7zip on Windows

In an effort to reduce external tools required to get started on Windows
remove use of 7zip in compression module now that we have a garuntee of
lzma support thanks to dropping support for Python < 3.7

johnwparent committed:
Style fixups

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -186,47 +182,47 @@ def _lzma_decomp(archive_file):
 
 
-def _win_compressed_tarball_handler(archive_file):
+def _win_compressed_tarball_handler(helper):
```
I think `decompressor` would be a clearer name than `helper`, since that's what it is doing.

johnwparent replied:
Good point! Will amend.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -186,47 +182,47 @@ def _lzma_decomp(archive_file):
     tool. Available only on Unix
     """
-    if sys.platform == "win32":
-        raise RuntimeError("XZ tool unavailable on Windows")
```
The docstring should be updated to reflect this change. The PR initial comment doesn't mention this but is `xz` availability a recent change on Windows?

johnwparent replied:
Yes it should, nice catch. XZ is still not available natively on Windows, however, this is more rectifying an inconsistency. Tar is the only tool in this compression module that is available natively on Windows. For unzip we shell out to tar, and for gz, bz2, and others if we can't use the python module for one reason or another we try to find something on the system and fail if nothing is found. This PR refactors the xz support to be the same as the others. Previously the xz python module wasn't even an option prior to Python3.4 (3.3?) so we had to have a fallback/default to 7zip as we just needed some way to extract Clingo. Now that we can more consistently (on Windows it should always be present) rely on the xz python module, we can refactor this to have the same fallback scheme as the rest. The end goal being not having a dependency on 7zip when spinning up Spack for the first time.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -186,47 +182,47 @@ def _lzma_decomp(archive_file):
-            return _7zip(archive_file)
-        else:
-            return _xz(archive_file)
+        return _xz(archive_file)
```
(question) You mention:

> Now that we've dropped support for these older version of Python, we can consistently rely on the lzma Python module existing to extract Clingo for installation.

but given that `xz` is now available on Windows, it wouldn't matter if there was an older version of Python, if I understand correctly.

johnwparent replied:
`xz`, like `gzip`, and `bz2` is not necessarily available on Windows now, this is just a consistency change. Ideally we're always going to use the python modules.

johnwparent committed:
REmove all references to 7zip

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -168,65 +168,57 @@ def _unzip(archive_file):
-    else:
-        result = _system_gunzip(archive_file)
-    return result
+    return _system_gunzip(archive_file)
```
Out of curiosity, why remove support for using 7zip? With the other changes (e.g. to `_win_compressed_tarball_handler`) we will default to using built-in Python support, but for `.Z` archives would it be problematic to fall back on `_7zip`?

johnwparent replied:
In general we should be trying to keep the number of dependencies needed to just get Spack running to a minimum, but I think in this case there's really no other option as there are no python options to extract `Z` files, and so long as we can install clingo, we can install 7zip with Spack, so I think it can be returned here. 
I know Spack loads all externally detected package prefixes into the `PATH` but is the same true with installed packages? Or would we need to manually make a spack installed 7zip available to the `which` method?

scheibelp replied:
> In general we should be trying to keep the number of dependencies needed to just get Spack running to a minimum,

Agree. But to be clear I don't think that the other changes in the PR already achieve that goal. For packages which happen to distribute `.Z`, providing the `_7zip` package would allow Spack to decompress them if the user installs 7zip.

> I know Spack loads all externally detected package prefixes into the PATH but is the same true with installed packages? 

That isn't automatic: we'd have to update our bootstrapping logic to also search for and include 7zip on Windows (but that could be done in some future PR, just allowing users to build `.Z`-based packages for now if they manually install 7zip is sufficient).

johnwparent replied:
This PR does not stop providing the the `_7zip` package from spack to users, it simply removes the dependence of the compression module on 7zip. Prior to this PR, it is impossible to use Spack without 7zip, as 7zip was required to extract the clingo compressed archive, which makes it a fundamental dependency, something we should avoid. 
This goal of this PR was to get 7zip to a place like `cmake` where we don't rely on it to get started but can install via Spack and use it from there. 

> That isn't automatic: we'd have to update our bootstrapping logic to also search for and include 7zip on Windows (but that could be done in some future PR, just allowing users to build .Z-based packages for now if they manually install 7zip is sufficient).

I had intended to mean loading a package installed via `spack install 7zip` rather than one that is bootstrapped. Agreed 7zip should probably be included in the bootstrap in a separate PR, but I'd strongly prefer to avoid requiring users to install packages themselves outside of Spack to use Spack, particularly for something so fundamental. So loading a spack installed version (if it exsits) would be ideal. Does that sound acceptable to you?

scheibelp replied:
> This PR does not stop providing the the _7zip package from spack to users,

I understand: there is still a `7zip` package.

However, if you remove the `_7zip` method here, how will spack ever be able to use `7zip` to unpack an archive?

> I'd strongly prefer to avoid requiring users to install packages themselves outside of Spack to use Spack, particularly for something so fundamental.

I agree and to be clear, keeping the `_7zip` method and offering it as an option for `.Z` files does not conflict with that goal.

johnwparent replied:
> However, if you remove the _7zip method here, how will spack ever be able to use 7zip to unpack an archive?

Sorry I thought I had conveyed that I agree with you in my initial comment, and that response was an attempt to clarify the need for this PR/how it changes the 7zip dep as I felt that maybe that wasn't as clear as I would like. For clarity I disagree with my initial implementation, agree with the above, and for `Z` compressed archives, we _should_ absolutely be falling back to 7zip as there is no other option that would not also put us in the same position. 

Is querying the installed packages (like `spack.store.db.query("7zip")[0].prefix.bin`) the best way to achieve this, or is there a more idiomatic way in Spack I'm unaware of? 
Obviously there would be more error handling in the actual implementation.

johnwparent committed:
7zipify Z compression

johnwparent committed:
First pass of compression refactor

Windows does not natively vendor any decompression utils, and only tar
for unpacking operations.
Windows should only try to use Python for decompression and native tar for
everything else.

Skip system unpacking test on Windows as that functionality is not
supported.

johnwparent committed:
style

johnwparent committed:
Update docstring

johnwparent committed:
Improve handling of abbreviated tarballs

johnwparent committed:
Style

johnwparent committed:
Merge branch 'develop' into win/un7zipify-codebase

johnwparent committed:
Correct tar archive check regex

johnwparent committed:
Merge branch 'win/un7zipify-codebase' of https://github.com/johnwparent/spack into win/un7zipify-codebase

johnwparent committed:
remove debug cruft

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -94,21 +94,36 @@ def _bunzip2(archive_file):
     working_dir = os.getcwd()
     archive_out = os.path.join(working_dir, decompressed_file)
-    copy_path = os.path.join(working_dir, compressed_file_name)
     if is_bz2_supported():
```
We should fail if this isn't true.

johnwparent replied:
Every other python decompression util will simply fail when the corresponding module is invoked if there is no support. Should we refactor all of them to raise a Spack style exception/error or allow this to fail as they do? Doesn't matter which direction we take IMO, just that they all fail the same way.

scheibelp replied:
> Every other python decompression util will simply fail when the corresponding module is invoked if there is no support. 

That works fine if this does that too. Right now it does not (it just returns archive_out as if it had decompressed the archive)

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -649,3 +745,37 @@ def check_and_remove_ext(path, ext):
+
+
+def reg_sub_ext(path, old_ext, new_ext):
+    return re.sub(rf"{old_ext}", rf"{new_ext}", path)
```
This is only used in one place, and since the thing it wraps is generally straightforward (and also one line) I think the method should be removed. If it stays, it should be indicated as private (leading `_` in name) and also I recommend a clearer name (e.g. `_substitute_extension`).

johnwparent replied:
I generally prefer to pull things out into methods if I can/they're a standalone concept, so I'll add the private method scoping and rename.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -615,6 +710,7 @@ def strip_last_extension(path):
 def strip_extension(path, ext=None):
     """Get the part of a path that does not include its compressed
     type extension."""
+    path = expand_contracted_extension_in_path(path)
```
`ALLOWED_ARCHIVE_TYPES` should include all possible abbreviations of extensions (e.g. `.tgz`) so I'm not clear what purpose this serves.

johnwparent replied:
I think I meant to put this logic in `strip_last_extension`

johnwparent replied:
Yep, looking at my notes I pretty clearly meant to put this in `strip_last_extension` and convert the decompression methods to use that instead of `strip_last_extension` like I did for the 7zip method. Amended this way locally.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -267,12 +290,14 @@ def _7zip(archive_file):
+
+def _validate_extension(path, extension=None):
     if not extension:
         extension = extension_from_file(path, decompress=True)
```
It seems strange to me to have a `validate...` method return a modified value. This is only called from `decompressor_for` so could you merge it back with that function?

johnwparent replied:
We're never returning a "modified" extension, we're returning the correct extension from the provided file path if the method is called without a given extension. I guess something like `_find_and_validate` would be more accurate, but sure, I can merge it back in.

scheibelp replied:
Sorry yes it is not modifying the value. I meant to say it is strange for a method named `...validate...` to extract a value

johnwparent replied:
Agreed, it was a confusing name.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -282,38 +307,108 @@ def decompressor_for(path, extension=None):
+        return _untar
+
+    extension = compression_ext_from_compressed_archive(extension)
+    decompressor = _determine_py_decomp_archive_strategy(extension)
```
A few things:

* Think this should only call `_determine_py_decomp_archive_strategy` once (just start out by expanding the extension if it's possibly a contraction like `.tgz`)
* Put the `if re.match(tar$, ext)` before that
* If `decompressor` is None at this point, we should raise an exception (this is likely an internal error, because we already would have detected if the extension is unsupported, and we aim to properly locate the appropriate tool for any extension that we do support).

johnwparent replied:
I disagree with some of this. If we perform this check a single time, we're either going to have to add anther check to determine if the extension we're looking at is a compressed tarball to decide whether to hand things off to `_win_compressed_tarball_handler` or a vanilla decompression method or we're just going to always blindly hand things off to that  windows compressed tarball method, which is conceptually incorrect and might result in confusing stack traces/is inconsistent with the rest of the changes suggested which are primarily along conceptual modeling lines. Either way we're making an "extra call" or modeling incorrectly.
The `tar` change is fine.

scheibelp replied:
I'm unclear where the following would go wrong:

```
if re.match(r"zip", extension):
    return _unzip

extension = compression_ext_from_compressed_archive(extension)
decompressor = _determine_py_decomp_archive_strategy(extension)
if decompressor:
    return decompressor

if re.match(r"tar$", extension):
    return _untar

raise Exception(...)
```

In other words, if `re.match(r"tar$", extension)` finds a match, I don't see how `_determine_py_decomp_archive_strategy(extension)` would return something (other than `None`).

If that still seems wrong, could you explain the expected results for `foo.tar`, `foo.tar.gz`, `foo.gzip`, and `foo.tbz`? Ultimately, regardless of how (or whether this changes) it might be useful to have that in the docstring of this function.

johnwparent replied:
Sorry, I was not clear. The `tar` check change is one of the parts I do agree with. It absolutely makes the most sense to check for and return that prior to checking for any type of decompression.
I also agree with the error reporting change.
My issue is with the 
> * Think this should only call _determine_py_decomp_archive_strategy once (just start out by expanding the extension if it's possibly a contraction like .tgz)


suggestion.


We need to call this twice. Once on the original extension, and once on the expanded extension if the original doesn't return anything useful. We _could_ only call this once, but in that case, we'd either need to perform a check to see if the extension is a compressed tarball or not, or to allow the `_win_compressed_tarball_handler` handle both normal vanilla decompression _and_ compressed tarballs, which, while the method is set up to do so for robustness reasons, is not really conceptually accurate, which seems like it would be a problem considering that many of the other suggestions here are along primarily conceptual modeling lines.

Edited for formatting.

scheibelp replied:
> We need to call this twice. Once on the original extension, and once on the expanded extension if the original doesn't return anything useful. 

I was thinking that if you start by calling `compression_ext_from_compressed_archive(extension)` to "normalize" it, then that would avoid having to call twice, but perhaps I misunderstood how that would work; I should say in that case that having such a normalizing function would be useful (which I assume would be simple - if that sounds like more than a few lines, I'm probably not understanding some underlying complexity).

johnwparent replied:
Right, so at that point in the code we know it's not a vanilla tarball, a zip file, or a vanilla .Z. We then start by "normalizing" the extension, so something like `tgz` becomes `tar.gz`, and things like `tar.gz` or `gz` stay the same. We then have a "normalized" ext where we don't have to think about the contraction extensions. We can then run the `_determine_py_decomp_archive_strategy(extension)` method on that extension, which gives us the compression type, `.gz` for all cases above. At this point, and this is where my issue arises, we would just be passing that derived decompressor to `_win_compressed_tarball_handler`, however the extension could be `gz` not `tar.gz` so passing it to the tarball handler as opposed to `_py_gunzip` doesn't really make a lot of conceptual sense, and in fact in addition to not making sense conceptually, without the conceptually redundant check for a tar extension being present I added to that handler, this approach would cause that to fail when tar tries to extract a file that is not a tarball. 
It's not really a normalization issue.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -649,3 +745,37 @@ def check_and_remove_ext(path, ext):
+
+
+def expand_contracted_extension(extension):
+    if extension in NOTAR_EXTS and extension != "zip":
```
Is `extension` always preceded by a `.` (i.e. is it a string of the form `r"\..*"`)? If so, I presume this check will always fail (i.e. `extension` could be `.zip` but never `zip`). If not, I think it would be good to wrap this in some consistency layer: strip all prefix `.` right away.

johnwparent replied:
No, `extension` is almost never preceded by a `.`, and the only way it could be is if the caller explicitly made this call with an extension prefixed with `.` (from `decompressor_for` or by calling this directly), but all other entrypoints from this module would guarantee there's no `.` in front of the prefix. Will add the `.` stripping.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -282,38 +307,108 @@ def decompressor_for(path, extension=None):
         return _gunzip
 
-    if re.match(r"bz2", extension):
+    if re.match(r"bz", extension):
```
I think this would always be `".bz2"` given how you have arranged things:

* `ALLOWED_ARCHIVE_TYPES` does not include `.bz`
* Your logic for expanding `.tbz` automatically converts to `.bz2`

I think though that this means we should add `"bz"` to the `ALLOWED_ARCHIVE_TYPES` (rather than look for "bz2" only here), since the logic here implies to me that there might be a `.tar.bz` file.

johnwparent replied:
`bz` and `bz2` are not the same, except for `tbz` which evidently is actually `bz2` in each case in Spack if my understanding is correct. If it's not, we need to revisit `ALLOWED_ARCHIVE_TYPES` from a conceptual standpoint as we currently do not support `bz` compression (unless bz2 happens to work for it?) and should then not support `tar.bz`, which is not something I intended to model. I think this was an early pass at a solution that it did not occur to me to refactor. 
So with that in mind I think you're right, this should always be `bz2` and at the moment, this change is effectively a big ole nothing burger as in the case of `tbz` this wouldn't match the prefixed `t`.

scheibelp replied:
OK I misunderstood, I thought you truncated from `bz2` to `bz` to accommodate both. If `.tbz` means `tar.bz2` then I think it would be clearer to just check for `"bz2$"`

johnwparent replied:
The convention I am familiar with says that a `tbz` is a `tar.bz2` but maybe that's not always the case if you're aware of an instance where its not (or any conventions/idioms/etc). I guess this is a good time to ask, do we want to support `bz`? I have no idea if bunzip/associated modules can handle that compression algo, so it would take some time research, but is certainly doable

scheibelp replied:
> I guess this is a good time to ask, do we want to support bz?

IMO no (at least not as part of this PR), although stricter regex matching would help if that gets added in the future.

johnwparent replied:
I've added the stricter matching. agreed on the non `bz`.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -282,38 +307,108 @@ def decompressor_for(path, extension=None):
     # we rely on external tools such as tar,
     # 7z, or uncompressZ
-    if re.match(r"Z$", extension):
+    if re.match(r"Z", extension):
```
Similar to the point about `bz`/`bz2`, I'm not clear how we can be doing anything for `".Z..."` if `allowed_archive` would return false for it: such extensions should be in `ALLOWED_ARCHIVE_TYPES`. This applies for any extension regex matching that doesn't include `$`.

johnwparent replied:
Yeah not sure why I took this out, but something like `.Z.<whatever other extension` also would not make it to this point in the code, but I'll make the change for conceptual modeling.

johnwparent replied:
We should, by this reasoning, always have a `$` at the end of each of the regex's, which is not something that was done prior to Spack on Windows, so I just want to confirm that this is correct.

scheibelp replied:
> We should, by this reasoning, always have a $ at the end of each of the regex's, which is not something that was done prior to Spack on Windows, so I just want to confirm that this is correct.

It is. I noticed the issue since this PR removes the trailing `$` in one case, which I assume is unnecessary; I agree with you it should always be fine (and probably clearer) to use `$` in the regexes

johnwparent replied:
Sounds good. already done locally.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -282,38 +307,108 @@ def decompressor_for(path, extension=None):
+    # we rely on 7zip, which must be installed outside
+    # of spack and added to the PATH or externally detected
+    if re.match(r"Z", extension):
+        return _unZ
```
I think this method should return `None` in this case and the caller should handle the consequences of that (as it already must for `.zip`): having it return a non-Python implementation in some cases is confusing (given the method's name).

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -168,6 +188,10 @@ def _unzip(archive_file):
 
 
 def _unZ(archive_file):
```
(minor) I think for clarity it might be worth calling this `_system_unZ`: I think if that were done, the naming conventions would be the following:

* `_system_x` uses a system utility
* `_py_x` uses builtin python support
* `_x` could use either

scheibelp reviewed (changes requested):
Broadly:

* I'm worried about a conceptual disconnect between `ALLOWED_ARCHIVE_TYPES` and some of the regular expressions
* I think some of the logic could be reorganized to give the various methods a clearer purpose

Let me know if you disagree.

johnwparent committed:
Review updates

johnwparent committed:
temp changes

johnwparent committed:
Cleanup after review

johnwparent committed:
Style

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -649,3 +741,44 @@ def check_and_remove_ext(path, ext):
+        # for compressed tarballs
+        # so we know tar is the first ext after initial '.'
+        prefix = "tar"
+        suffix = extension[1:]
```
If extension is a string like "tar", it looks like this would be problematic (part of why I asked about preceding "." above was because I couldn't think of another reason for stripping off the first item of extension). Could you also explicitly say in the docstring that extension is a string (i.e. "tar.gz" rather than a list like `["tar", "gz"]`

johnwparent replied:
I'm always happy to improve a docstring, will do. 
With regard to the tar concern, maybe we should discuss at the meeting tomorrow, as I'm a bit unsure how this could be problematic. For example, I just tried this out locally:
```
[spack] C:\spack_win\spack>spack python
Spack version 0.20.0.dev0
Python 3.8.4, Windows AMD64
>>> import spack.util.compression as sc
>>> sc.expand_contracted_extension("tar")
'tar'
>>> sc.expand_contracted_extension(".tar")
'tar'
>>> sc.expand_contracted_extension(".tar.gz")
'tar.gz'
>>> sc.expand_contracted_extension(".tgz")
'tar.gz'
>>> sc.expand_contracted_extension(".tar.bz2")
'tar.bz2'
>>> sc.expand_contracted_extension(".tbz2")
'tar.bz2'
>>> sc.expand_contracted_extension(".tbz")
'tar.bz2'
```
Which seems to be behaving as expected.

johnwparent replied:
I also don't believe an extension "tar" would make it to that code block as the preceding if statement would preclude that. At that point in execution, we would know that the "extension" is one of `"tgz", "txz", "tbz2", "tbz"` so stripping off the first character (the t) would give us the true extension (with the exception of `tbz` which I handle later).

scheibelp replied:
OK I see that all `NOTAR_EXTS` except zip/tbz would be handled this way. In my case I didn't understand how this worked until you pointed that out. IMO a dictionary like {"tbz": "bz2", "tgz": "tar.gz"...} would be more straightforward (and also possibly avoid future issues if more entries are added to `NOTAR_EXTS`).

johnwparent replied:
I've been avoiding taking that approach as I assumed that a hardcoded dictionary supplementary to `NOTAR_EXTS` would just be an added and easy to miss maintenance burden. If not, I can add it.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -602,13 +691,16 @@ def extension_from_path(path):
-            mod_path = check_and_remove_ext(path, ext)
-            if mod_path != path:
-                return mod_path
+def strip_last_extension(path, ext=None):
```
I think `strip_compression_extension` would be a clearer name (IMO a function called `strip_last_extension` would transform `foo.tgz -> foo`).

johnwparent replied:
Good point! Will amend.

johnwparent committed:
Review updates

johnwparent committed:
Style

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -282,38 +299,111 @@ def decompressor_for(path, extension=None):
+
+    # if extension is standard tarball, invoke Windows native tar
+    if re.match(r"tar$", extension):
+        return _untar
```
Could you rename `def _untar` to `def _system_untar` and update its docstring? It looks out of date.

johnwparent replied:
Sure

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -167,7 +187,11 @@ def _unzip(archive_file):
+    Utilizes gunzip on unix and 7zip on Windows
+    """
     if sys.platform == "win32":
         result = _7zip(archive_file)
```
Can you rename `_7zip` to `_system_7zip`.

johnwparent replied:
IMO the system part of 7zip should be implied considering no python module exists for it, but I'll make the change anyway.

scheibelp replied:
Thanks: I understand there is no module for it, but I think it is worthwhile to make it obvious with the method name.

johnwparent replied:
Very reasonable. Additionally I appear to have put my foot in my mouth as it turns out someone did actually make a 7zip python module and put it on pypi. Not a standard lib, but I'm still very incorrect.

scheibelp reviewed (changes requested):
I have a couple additional requests along the lines of https://github.com/spack/spack/pull/36507#discussion_r1171953613

johnwparent committed:
Elucidate function names

johnwparent committed:
Style

johnwparent committed:
Check for tar in full extension not comp ext

johnwparent committed:
Decouple ext expansion from ext stripping

scheibelp commented:
@spackbot run pipeline
